### ISSUE TITLE
@SOURCE_ROOT@ doesn't work in custom_target

### ISSUE THREAD

**Issue by Hi-Angel** [mesonbuild/meson#9212]
**Describe the bug**

[Documentation says](https://mesonbuild.com/Reference-manual.html#custom_target), quoting:

> The list of strings passed to the command keyword argument accept the following special string substitutions:
> […]
> * @SOURCE_ROOT@: the path to the root of the source tree. Depending on the backend, this may be an absolute or a relative to current workdir path

However, trying to use `@SOURCE_ROOT@` result in `Program '@SOURCE_ROOT@' not found`

**To Reproduce *(in terms of terminal commands)***

```
 λ cat meson.build
project(
  'mesa',
  ['c'],
)

custom_target(
  'foo',
  input : 'foo.sh',
  command : ['@SOURCE_ROOT@', '/foo.sh']
)
 λ touch foo.sh && chmod +x foo.sh
 λ meson build
The Meson build system
Version: 0.59.1
Source dir: /tmp
Build dir: /tmp/build
Build type: native build
Project name: mesa
Project version: undefined
C compiler for the host machine: ccache cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program @SOURCE_ROOT@ found: NO

meson.build:6:0: ERROR: Program '@SOURCE_ROOT@' not found

A full log can be found at /tmp/build/meson-logs/meson-log.txt
```

**Expected behavior**

There is no complaint about `@SOURCE_ROOT@` being an unknown command

**Actual behavior**

There is complaint about `@SOURCE_ROOT@` being an unknown command

**system parameters**
* build type: native
* OS: Archlinux
* python: 3.9.6
* `meson --version`: 0.59.1
* `ninja --version`: 1.10.2

**Comment by Hi-Angel** [comment]
FTR, the workaround here is adding a line `foo_script = files('foo.sh')`, for example:

```
…
foo_script = files('foo.sh')

custom_target(
  'foo',
  input : foo_script,
  command : [foo_script, '/foo.sh']
)
```

**Comment by eli-schwartz** [comment]
It does work in any position other than the command name, e.g. `command: ['echo', '@SOURCE_ROOT@/foo.sh']`

I wonder if maybe this is just a documentation issue and we should say these replacements are only valid as arguments in the command, but not as the first element.

P.S. The correct workaround is to use `foo_script = find_program('foo.sh')`. Note that recent versions of meson automatically run find_program on the first element of the command, if it is a simple string.

**Comment by JamBalaya56562** [comment]
To note the current state, having re-checked this end-to-end against current master:

**The behavior is by-design, and this is not a bug in the substitution engine.**

The `command:` array of a `custom_target()` is treated in two different ways:

- The **first element** (the program to run) is resolved via `find_program()` *at interpret time*, before any `@…@` templates are expanded. So a literal `'@SOURCE_ROOT@'` in position 0 is looked up as if it were a program name, which is why you get `Program '@SOURCE_ROOT@' not found`:

```python
# mesonbuild/interpreter/interpreter.py  (custom_target creation)
command = kwargs['command']
if command and isinstance(command[0], str):
    command[0] = self.find_program_impl([command[0]])
```

- The `@SOURCE_ROOT@` / `@BUILD_ROOT@` / `@CURRENT_SOURCE_DIR@` / … substitutions are applied *later, in the backend*, and only to elements that are still plain strings (i.e. the **arguments**). By the time the backend runs, the first element is already a resolved `Program` object, so it is passed through untouched and never substituted:

```python
# mesonbuild/backend/backends.py  (eval_custom_target_command)
elif isinstance(i, str):
    if '@SOURCE_ROOT@' in i:
        i = i.replace('@SOURCE_ROOT@', source_root)
    ...
elif isinstance(i, programs.Program):
    # Let it pass and be extended elsewhere
    pass
```

So, as @eli-schwartz noted in the thread:

- `@SOURCE_ROOT@` **does** work correctly as an *argument*, e.g. `command : ['echo', '@SOURCE_ROOT@/foo.sh']`.
- Using it as the *command name* (position 0) is not supported and never was.
- The supported way to point at a script in the source tree is to resolve it explicitly, e.g. `command : [find_program('foo.sh')]` or `command : [files('foo.sh')]`.

**What is resolved:** the underlying behavior is intentional and unchanged, and a supported/documented mechanism (`find_program`/`files` for the program element; `@SOURCE_ROOT@` for arguments) covers the reporter's use case.

**What is still open:** the concrete follow-up this issue converged on — @eli-schwartz's suggestion to *document* that these replacements are only valid for command arguments, not for the first (program) element — does not appear to have been done. The [`custom_target` reference](https://mesonbuild.com/Reference-manual.html#custom_target) and `docs/yaml/functions/custom_target.yaml` still list the substitutions without any note that they don't apply to the command name, and `Custom-build-targets.md` doesn't mention it either. So the documentation-clarity part of this report is genuinely outstanding.

I'll leave the disposition to the maintainers — it may be worth keeping open as a small docs task, or closing if the docs note is added.

---

### Verification

- Checked against current master (checkout `1a7522f1d0db`, GitHub master `2b4f5dcd1b37`).
- Program-element resolution: `mesonbuild/interpreter/interpreter.py`, `custom_target` handling — `command[0]` is passed through `find_program_impl()` before substitution.
- Argument-only substitution: `mesonbuild/backend/backends.py`, `eval_custom_target_command()` — `@SOURCE_ROOT@` etc. are replaced only on `str` elements; a resolved `programs.Program` (the first element) is passed through unchanged. The substitution loop itself dates back to commit `1b7855c36d6e` ("Handle multiple substitutions in eval_custom_target_command()", 2020-06-13) and is unchanged in intent.
- Docs check: `docs/yaml/functions/custom_target.yaml` (rendered at https://mesonbuild.com/Reference-manual.html#custom_target) lists the `@SOURCE_ROOT@`/`@BUILD_ROOT@`/`@CURRENT_SOURCE_DIR@` substitutions but contains no caveat that they do not apply to the first/program element — i.e. the requested clarification is not yet present.

**Comment by Hi-Angel** [comment]
@JamBalaya56562 am I missing something? Your text looks like a much more elaborate reiteration of the 3 comments already present in the thread, am I missing a point perhaps?

**Comment by JamBalaya56562** [comment]
You're right — most of that just restated the existing explanation, sorry for the noise. The only substantive point I meant to add is that the docs (the `custom_target` reference and `Custom-build-targets.md`) still don't mention that these `@…@` substitutions don't apply to the command's first (program) element, so this might be worth keeping open as a small docs task.

**Comment by bonzini** [comment]
Hi @JamBalaya56562 , as a rule do *not* copy and paste comments to the LLM and responses back. That is something everyone agrees we absolutely do not want in the Meson project. Please leave further communication to the maintainers and limit yourself to verifying and including the original triaging.

Generally speaking I would consider the posting of almost a hundred comments to be borderline acceptable. In this specific case, it has been certainly and improvement to the state of the meson issue tracker, and that's why I am accepting it, but not everyone might agree with me.

Please stick to some ground rules when submitting your triage. In addition to the above request please limit your submissions to a few comments per day to avoid burning out the maintainers; or, if you have an up to date document with all the results of the automated triaging, you can link it and let maintainers process it at their own leisure.

**Comment by JamBalaya56562** [comment]
Hi @bonzini — thank you for the fair and measured response, and I'm sorry for the noise.

I take the ground rules to heart. I started the triage because I found existing issues hard to search — many were already resolved but left open — and that got cleaned up thanks to the maintainers' work, which I appreciate. I'll stop commenting on other issues from now on.

Since this is off-topic here and needs no reply, feel free to delete it once read. Thanks again, and sorry for the trouble.

**Comment by bonzini** [comment]
> I'll stop commenting on other issues from now on.

You can comment on dormant issues that you think can be closed, I will take responsibility for following up on those. Just do so with moderation. And again, if you have the results of your triaging somewhere in a file you can open a single issue that links to that.
